### PR TITLE
Make slice generic

### LINKED ISSUE #8647: `builtins.slice` should be generic
This is a port of https://github.com/python/typing/issues/159, which was closed as being out of scope, with https://github.com/srittau suggesting it maybe be more suitable for [typeshed](https://github.com/python/typeshed) instead.

It would be useful to have `slice` being generic and allowing type hints a la

- `slice[start_type, stop_type, step_type]`
- `slice[S, T]` as shorthand for `slice[S, S, T]` or `slice[S, T, Any]`
- `slice[T]` as shorthand for `slice[T, T, T]` or `slice[T, T, Any]` or `slice[Any, T, Any]`
- `slice` as shorthand for `slice[Any, Any, Any]`

Syntax wise, I would advocate for the latter options `slice[S, T]=slice[S, T, Any]`, `slice[T]=slice[Any, T, Any]` for the following reasons:

1. It resembles the regular usage of `slice`: 
    - `slice(x) = slice(None, x, None)`, 
    - `slice(a, b) = slice(a, b, None)`
2. It has better compatible with `datetime`/`timedelta` variables, which might be a common use-case.
   - `slice[datetime]=slice[datetime, datetime, datetime]` would be an error, because `datetimes cannot be added, only timedeltas can be added to datetimes.

### Real-world examples

- Datetime based slicing: Here, `start` and `stop` are `datetime`-like and `step` is `timedelta`-like (e.g. `datetime.datetime` / `datetime.timedelta`, `numpy.datetime64`/`numpy.timedelta64` or `pandas.Timestamp`/`pandas.Timedelta`)
- Label based slicing via strings, for example, `pandas` allows something like `frame.loc["2017-07-01":"2017-10-31"]` for a `DataFrame` that is indexed by a `DatetimeIndex`, this will first convert the strings to `datetime` objects and then select all rows of the table that lie between the given dates.
- Similarly, one might want to select all rows of a table that lie between two float values.

### PR DESCRIPTION
**Pull request by Sachaa-Thanasius**
Mainly doing this out of curiosity to see the primer hits.

EDIT: Would close #8647.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
cc @cdce8p I think the primer hits suggest a mypy bug where it looks up "_StartT" in the wrong scope.

**Comment by cdce8p** [comment]
> cc @cdce8p I think the primer hits suggest a mypy bug where it looks up "_StartT" in the wrong scope.

Recursive TypeVar defaults was one of the things that didn't make it into `1.9.0` unfortunately. I did do some work on it already, but that isn't released yet: https://github.com/python/mypy/pull/16878. Don't know if it's enough for `slice` yet though.

If you're interested, maybe that's also something for @AlexWaygood, a few weeks back I created a custom repo to upload dev releases of mypy to PyPI under the [mypy-dev](https://pypi.org/project/mypy-dev/) project name. They always track a mypy commit upstream and are great to test / use the lasted version (even in production if you feel like it). _`mypy --version` still works as expected._ I originally created it as it was unclear when `1.9.0` would actually be released. In the past the release frequency was also a bit more unpredictable.

https://github.com/cdce8p/mypy-dev
https://github.com/cdce8p/mypy-dev/releases

**Comment by Sachaa-Thanasius** [comment]
I just assumed my attempt at implementation was wrong and gave up. Should this be reopened, then?

**Comment by JelleZijlstra** [comment]
Thanks @cdce8p, I forgot that PR didn't make it into 1.9. Hopefully we'll have another release soon. Your mypy-dev project is interesting; maybe that's even something we could do on mypy itself. For example, we could have a weekly cron job uploading alpha releases.

@Sachaa-Thanasius I think we may be able to merge this PR after mypy 1.10 is released (when that will happen, I don't know). You can leave it open in the meantime and mark it as deferred. With some hackery it may be possible to point typeshed's CI at the mypy master branch so we can test earlier whether everything works as expected.

**Comment by Sachaa-Thanasius** [comment]
Understood. Should slice be changed in the runtime cpython as well to support `__class_getitem__` via `GenericAlias`, or is this purely meant to be a typeshed thing? Not something I considered when originally opening this, but now that this actually has a chance of going somewhere . . .

**Comment by Sachaa-Thanasius** [comment]
Also, sidenote: The pyright errors specifically for `slice.stop` and and the first overload of `slice.__new__` (with text like the following) are the result of a bug that was fixed [earlier today](https://github.com/microsoft/pyright/issues/7525), I think:
```
Type parameter "_StopT" has a default type that refers to one or more type variables that are out of scope
    Type variable "_StartT" is not in scope
```

**Comment by cdce8p** [comment]
> With some hackery it may be possible to point typeshed's CI at the mypy master branch so we can test earlier whether everything works as expected.

@Sachaa-Thanasius You could try to replace the mypy version listed in `requirements-tests.txt`.
https://github.com/python/typeshed/blob/5c75292f26aec8e6348a17d32077f408a1337988/requirements-tests.txt#L8
```diff
-mypy==1.9.0
+mypy-dev==1.10.0a2
```

**Comment by cdce8p** [comment]
> > With some hackery it may be possible to point typeshed's CI at the mypy master branch so we can test earlier whether everything works as expected.
> 
> @Sachaa-Thanasius You could try to replace the mypy version listed in `requirements-tests.txt`.
> 
> https://github.com/python/typeshed/blob/5c75292f26aec8e6348a17d32077f408a1337988/requirements-tests.txt#L8
> 
> ```diff
> -mypy==1.9.0
> +mypy-dev==1.10.0a2
> ```

_That release tracks https://github.com/python/mypy/commit/b013cc016a3e3c8c7caa6a27bdf7b5e22998dd41 which was commited fairly recently to mypy master (5 days ago)._

**Comment by Sachaa-Thanasius** [comment]
Well, that's far more promising than before.

**Comment by srittau** [comment]
I've removed the Deferred label. I've just looked at spark primer output. The problem is on their side, since they reuse the `start` variable for different types.

**Comment by Sachaa-Thanasius** [comment]
The error that pyright is hitting is that the usage of `itertools.starmap` doesn't resolve the types of `slice`.

```python
def subslices(seq: Sequence[_T]) -> Iterator[Sequence[_T]]:
    slices = starmap(slice, combinations(range(len(seq) + 1), 2))
    reveal_type(slices)  # Type of "slices" is "starmap[slice[_StartT@slice, _StopT@slice, _StepT@slice]]"
    # Adding this cast makes the error go away:
    # slices = cast(starmap[slice[int, int, int | None]], slices)
    return map(getitem, repeat(seq), slices)  # Long error.
```

Might be more of a deficiency in the typing of `starmap` than anything with `slice`.

**Comment by hamdanal** [comment]
I might be missing something obvious but it is not clear to me why the type of `stop` has to default to the type of `start`. At runtime `start` is `None` if it is not passed to the constructor:

```console
$ cat t.py
from typing_extensions import reveal_type

s = slice(10)
reveal_type(s)
reveal_type(s.start)
reveal_type(s.stop)
reveal_type(s.step)
$ python t.py
Runtime type is 'slice'
Runtime type is 'NoneType'
Runtime type is 'int'
Runtime type is 'NoneType'
```
Similarly, the type of `step` should default to `None`.

So I would expect the definitions of the type variables to be:
```python
_StartT = TypeVar("_StartT", default=None)
_StopT = TypeVar("_StopT", default=int)
_StepT = TypeVar("_StepT", default=None)
```

**Comment by Sachaa-Thanasius** [comment]
I won’t lie; I just copy-pasted the typeshed implementation from PEP 696 without much thought, since it provided one and this PR wasn’t originally opened with the intention of being merged. Haven’t double-checked since. Definitely will take another look though.

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -91,6 +91,9 @@ _SupportsAnextT = TypeVar("_SupportsAnextT", bound=SupportsAnext[Any], covariant
 _AwaitableT = TypeVar("_AwaitableT", bound=Awaitable[Any])
 _AwaitableT_co = TypeVar("_AwaitableT_co", bound=Awaitable[Any], covariant=True)
 _P = ParamSpec("_P")
+_StartT = TypeVar("_StartT", default=int)
+_StopT = TypeVar("_StopT", default=_StartT)
+_StepT = TypeVar("_StepT", default=int | None)
```
Comment (srittau) [inline stdlib/builtins.pyi:96]: Maybe as a very first step, try using `Any` as default. We could change this later, but this should allow explicit annotations without regressions.

**Comment by Sachaa-Thanasius** [comment]
Maybe I'm misreading, but some of these errors indicate that mypy is using the typevar defaults as the typevar bounds as well, which isn't correct afaik.

**Comment by hamdanal** [comment]
I'd go with @srittau's idea of setting defaults to `Any` https://github.com/python/typeshed/pull/11637#discussion_r1596543856. This will allow users to parametrize `slice` without any regressions and we can always change the defaults later if needed.

**Comment by hamdanal** [comment]
One more thing, shouldn't the type variables be covariant?

**Comment by Sachaa-Thanasius** [comment]
> One more thing, shouldn't the type variables be covariant?

I feel like I don't know enough about how slices are used beyond the basics to know, tbh.

**Comment by JelleZijlstra** [comment]
Determining variance doesn't really require knowing how the class is used. Roughly speaking, a type parameter can be covariant if it's only used in method return types and contravariant if it's only used in parameter types. Here the typevars are only used in return types (read-only properties), so they can be covariant.

Type checkers will tell you if you use a covariant or contravariant typevar where you should have used an invariant one.

**Comment by srittau** [comment]
Covariance might help with some of the primer hits. I also don't understand why mypy wants explicit annotations in the test cases.

**Comment by srittau** [comment]
Looking at the primer hits:

1. pandas does variations of this:

    ```python
      axis_indexer = [slice(None)] * values.ndim
      if periods > 0:
          axis_indexer[axis] = slice(None, periods)
      else:
          axis_indexer[axis] = slice(periods, None)
    ```

    This will need an explicit annotation of `axis_indexer`. Not great, but not unheard of for lists. I also expect that collecting slices into a list is a fairly uncommon use, outside of pandas.
2. kornia is collecting the slice in a variable, before using it. Will also need an explicit annotation.
3. jax hit number 1 is a bit more problematic:
  
    ```python
    slices = tuple(slice(int(d._data)) if type(d) is DArray and
                       type(d.dtype) is bint else slice(None) for d in self.shape)
    ```
4. The other jax hits need explicit annotations for lists.

While this introduces a bit of churn in more complex/unorthodox uses of `slice`, I'm +1 on this change if we can figure out why the mypy tests fail.

**Comment by hamdanal** [comment]
I don't know why the mypy test is failing but I figured out how to make it work. If we decouple the class scoped type variable from the `__new__` method scoped type variables mypy will be able to figure out how to handle them and the failing test passes. I don't see a downside to returning `slice` instead of `Self` as the class is marked with final. Here is a patch:
```patch
diff --git a/stdlib/builtins.pyi b/stdlib/builtins.pyi
index 6e26bab06..bcc44c0c5 100644
--- a/stdlib/builtins.pyi
+++ b/stdlib/builtins.pyi
@@ -75,6 +75,7 @@ if sys.version_info >= (3, 9):
     from types import GenericAlias

 _T = TypeVar("_T")
+_R = TypeVar("_R")
 _T_co = TypeVar("_T_co", covariant=True)
 _T_contra = TypeVar("_T_contra", contravariant=True)
 _R_co = TypeVar("_R_co", covariant=True)
@@ -925,9 +926,11 @@ class slice(Generic[_StartT, _StopT, _StepT]):
     @property
     def stop(self) -> _StopT: ...
     @overload
-    def __new__(cls, stop: _StopT, /) -> Self: ...
+    def __new__(cls, stop: _S, /) -> slice[Any, _S]: ...
     @overload
-    def __new__(cls, start: _StartT, stop: _StopT, step: _StepT = ..., /) -> Self: ...
+    def __new__(cls, start: _R, stop: _S, /) -> slice[_R, _S]: ...
+    @overload
+    def __new__(cls, start: _R, stop: _S, step: _T, /) -> slice[_R, _S, _T]: ...
     def __eq__(self, value: object, /) -> bool: ...
     __hash__: ClassVar[None]  # type: ignore[assignment]
     def indices(self, len: SupportsIndex, /) -> tuple[int, int, int]: ...
```

**Comment by AlexWaygood** [comment]
I fixed the failing test using @hamdanal's excellent suggestion and merged in `main`. Now it's just a matter of figuring out whether the new primer hits are acceptable or not. @srittau analysed a bunch of these in https://github.com/python/typeshed/pull/11637#issuecomment-2149504790, but there seem to be some new ones now. `freqtrade` and `xarray` aren't too concerning; these are places where users will have to add explicit annotations, but it's not so awful. These are a bit concerning to me, however:

> ```diff
> pandas-stubs (https://github.com/pandas-dev/pandas-stubs)
> + tests/test_frame.py:224: error: Expression is of type "Any", not "DataFrame"  [assert-type]
> + tests/test_frame.py:225: error: Expression is of type "Any", not "DataFrame"  [assert-type]
> ```

The failing `assert_type` calls are [here](https://github.com/pandas-dev/pandas-stubs/blob/f71224c6211ff436424be614ba47f494f64c4618/tests/test_frame.py#L224-L225). I'm not sure I understand why mypy is now inferring `Any` for `df[1:]`, as this should be covered by the second overload for `DataFrame.__getitem__` [here](https://github.com/pandas-dev/pandas-stubs/blob/f71224c6211ff436424be614ba47f494f64c4618/pandas-stubs/core/frame.pyi#L605-L610), which accepts a union including `slice` and returns `DataFrame`.

**Comment by JukkaL** [comment]
If you can narrow down some false positives to a mypy bug, it would help if you can file a mypy issue and include a link to this PR.

**Comment by AlexWaygood** [comment]
The `pandas-stubs` errors I discussed in https://github.com/python/typeshed/pull/11637#issuecomment-2435153478 are now the _only_ new false positives this PR introduces. I would say the cost-benefit is now in this PR's favour. (For one thing, we'd possibly find it quite useful for red-knot.)

@JukkaL, I'll see if I can narrow down the strange `pandas-stubs` primer hits to a mypy bug.

**Review by srittau — APPROVED** [review]
We can always refine this later, this is a good start!

**Review by AlexWaygood — APPROVED** [review]

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -936,17 +940,25 @@ class bool(int):
...
+    def start(self) -> _StartT: ...
     @property
-    def step(self) -> Any: ...
+    def step(self) -> _StepT: ...
     @property
-    def stop(self) -> Any: ...
+    def stop(self) -> _StopT: ...
     @overload
-    def __new__(cls, stop: Any, /) -> Self: ...
+    def __new__(cls, stop: int | None) -> slice[int | MaybeNone, int | MaybeNone, int | MaybeNone]: ...
```
Comment (randolf-scholz) [inline stdlib/builtins.pyi:951]: changed positional-only argument to positional or keyword.

Comment (randolf-scholz) [reply stdlib/builtins.pyi:951]: slice takes no keyword arguments

Comment (randolf-scholz) [reply stdlib/builtins.pyi:951]: Fixed by #12900

**Comment by AlexWaygood** [comment]
> @JukkaL, I'll see if I can narrow down the strange `pandas-stubs` primer hits to a mypy bug.

@JukkaL I finally found some time to minimize this: https://github.com/python/mypy/issues/18149

**Pull request merged**